File: Problem/Yutae/Week1/B1074(Z).cpp
// https://www.acmicpc.net/problem/1074

#include <iostream>
#include <cmath>

using namespace std;

int N, r, c;
int num[4];

int main() {
	cin >> N >> r >> c;

	int start = 0;
	while (1) {
		for (int i = 0; i < 4; i++)
			num[i] = i * pow(2, 2 * N - 2) + start;

		if (N == 1) break;

		int half = pow(2, N - 1);
		if (r < half && c < half)
			start = num[0];
		else if (r < half && c >= half) {
			start = num[1];
			c -= half;
		}
		else if (r >= half && c < half) {
			start = num[2];
			r -= half;
		}
		else if (r >= half && c >= half) {
			start = num[3];
			c -= half;
			r -= half;
		}
		N--;
	}

	cout << num[2 * r + c] << "\n";


	return 0;
}

File: Problem/Sehyeon/Week1/B1074(Z).java
https://www.acmicpc.net/problem/1074

import java.util.Scanner;

public class B1074 {
	static int N;
	static int R;
	static int C;
	static int answer;
	
	static int findSection(int halfSideSize, int y, int x) {
		int section = 0;
		
		if (R >= halfSideSize + y)
			section += 2;
		if (C >= halfSideSize + x)
			section += 1;
		return (section);
	}
	
	static void divideConquer(int sideSize, int y, int x) {
		int section;
		int oneSectionSize;
		int halfSideSize;
		
		if (sideSize == 1)		
			return ;
		
		halfSideSize = sideSize / 2;
		section = findSection(halfSideSize, y, x);
		oneSectionSize = (int)Math.pow(halfSideSize, 2);
		
		answer += (oneSectionSize * section);
		if (section == 1)
			x += halfSideSize;
		else if (section == 2)
			y += halfSideSize;
		else if (section == 3) {
			y += halfSideSize;
			x += halfSideSize;	
		}
		divideConquer(halfSideSize, y, x);
	}
	
	static void solution() {
		divideConquer((int)Math.pow(2, N), 0, 0);
	}
	
	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);
		
		N = scan.nextInt();
		R = scan.nextInt();
		C = scan.nextInt();
		
		answer = 0;
		solution();
		System.out.println(answer);
	}
}
